query/dev/system/src/main/java/org/codice/ddf/admin/query/dev/system/persist/CreateFeatureDependencyGraph.java

```diff
@@ -0,0 +1,143 @@
+import java.util.stream.Collectors;
+
+import org.apache.karaf.features.Feature;
+import org.codice.ddf.admin.api.Field;
+import org.codice.ddf.admin.api.fields.FunctionField;
+import org.codice.ddf.admin.common.fields.base.BaseFunctionField;
+import org.codice.ddf.admin.common.fields.base.scalar.BooleanField;
+import org.codice.ddf.admin.common.fields.base.scalar.StringField;
+import org.codice.ddf.admin.query.dev.system.dependency.FeatureUtils;
+import org.codice.ddf.admin.query.dev.system.fields.FeatureField;
+import org.codice.ddf.admin.query.dev.system.graph.FeatureGraphProvider;
+import org.codice.ddf.admin.query.dev.system.graph.DependencyEdge;
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.ext.ExportException;
+import org.jgrapht.ext.GraphMLExporter;
+import org.jgrapht.graph.DirectedPseudograph;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+public class CreateFeatureDependencyGraph extends BaseFunctionField<BooleanField> {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(CreateFeatureDependencyGraph.class);
+
+  private static final String DEFAULT_GRAPH_NAME = "featureDependenciesGraph.graphml";
+  private static final String FUNCTION_NAME = "createFeatureGraph";
+  private static final String DESCRIPTION = "Saves a graph called \'" + DEFAULT_GRAPH_NAME + "\' consisting of bundles and features as the vertices."
+          + "By default, the graph save path is under ddf.home. The file format is in graphml. Look here for more information: http://graphml.graphdrawing.org/";
+
+  private static final BooleanField RETURN_TYPE = new BooleanField();
+  private static final Set<String> ERROR_CODES = ImmutableSet.of(DIRECTORY_DOES_NOT_EXIST);
+  private static final String SAVE_DIR_FIELD_NAME = "saveDirectory";
+
+  private StringField saveDir;
+  private List<Field> arguments;
+
+  private static final FeatureGraphProvider.FeatureOrBundleVertexAttributeProvider
+      FEATURE_BUNDLE_VERTEX_PROV =
+          new FeatureGraphProvider.FeatureOrBundleVertexAttributeProvider();
+  private FeatureUtils featureUtils;
+
+  // Leaving nodes as Field since they can either be a FeatureField or a BundleField
+  private GraphMLExporter<Field, DependencyEdge> exporter;
+
+  public CreateFeatureDependencyGraph(FeatureUtils featureUtils) {
+    super(FUNCTION_NAME, DESCRIPTION);
+    saveDir = new StringField(SAVE_DIR_FIELD_NAME);
+    arguments = ImmutableList.of(saveDir);
+
+    this.featureUtils = featureUtils;
+    exporter = new GraphMLExporter<>();
+    exporter.setVertexAttributeProvider(FEATURE_BUNDLE_VERTEX_PROV);
+  }
+
+  @Override
+  public BooleanField performFunction() {
+    DirectedGraph<Field, DependencyEdge> graph =
+        createFeatureDependenciesGraph(featureUtils.getAllFeatures());
+    String savePath =
+        saveDir.getValue() == null ? System.getProperty("ddf.home") : saveDir.getValue();
```

Comment:
Should this saveDir be validated? For instance, checking access and existence.

Reply:
I created a new field "DirectoryField" and added optional validation to check if the directory exists. All the functions set this.